### PR TITLE
[FLINK-16177][checkpointing] Integrate OperatorCoordinator checkpoint triggering and committing

### PR DESCRIPTION
**Pull request by StephanEwen**
## What is the purpose of the change

**This is part of [FLIP-27](https://cwiki.apache.org/confluence/display/FLINK/FLIP-27%3A+Refactor+Source+Interface?src=contextnavpagetreemode)**

This integrates the previously added `OperatorCoordinators` with checkpointing, in particular with checkpoint triggering and committing.

This PR implements initially a simplified version of checkpoint triggering that does not offer the facilities to align coordinator checkpoints and source checkpoint barrier triggering. 

## Brief change log

  - Add "Operator Coordinator State" to the "OperatorState"
  - Introduce a new version of the Checkpoint Metadata format
  - Add Coordinator State to `PendingCheckpoint`
  - Add a new stage to the checkpoint triggering procedure

## Not Addressed issues

Separate follow-ups are needed for the following two issues:

  - **Checkpoint restoring** The operator coordinators only need to restore on "global restore", meaning when the master fails over, or when a "global job failure" is triggered. Otherwise, the coordinators are notified of local task failures (for example to invalidate specific split assignments).

    To support that, the calls from the scheduler to the checkpoint coordinator need to differentiate between "common failover" (pipelined region) and full job failover (safety net, when encountering inconsistencies / unexpected exceptions in the execution graph or in the coordinators). See [FLINK-16357](https://issues.apache.org/jira/browse/FLINK-16357)

  - **Exactly Once** alignment between coordinator checkpoints and source barrier injection is not included. In this version here, the source barriers are injected once all coordinators are done with their checkpoints. There exist different ideas for how to deal with in flight `OperatorEvents`, which we plan to validate when implementing the `SourceEnumeratorCoordinator`. Hence, this is not yet included, but the code is written to support adding these changes isolated (in the `OperatorCoordinatorCheckpointContext`).

## Verifying this change

This change is already covered by existing tests, and partially extends the class `PendingCheckpointTest`.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): **no**
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: **no**
  - The serializers: **no**
  - The runtime per-record code paths (performance sensitive): **no**
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: **yes**
  - The S3 file system connector: **no**

## Documentation

  - Does this pull request introduce a new feature? **no**
  - If yes, how is the feature documented? **not applicable**

### CODE REVIEW

File: flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/DisposableAsync.java [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+ */
+public interface DisposableAsync {
+
+	/**
+	 * Disposes this object.
+	 *
+	 * <p>The given executor may be used to schedule expensive/blocking operations where it is
+	 * acceptable that they complete only later.
+	 */
+	void dispose(Executor executor) throws Exception;
```
Comment (becketqin) [inline flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/DisposableAsync.java]: It seems this method is not called anywhere in the patch. Is this intended to be used by the coordinator implementations?

Comment (StephanEwen) [reply flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/DisposableAsync.java]: This file snuck in. It is part of another follow-up patch (simplifying the trigger procedure). Will remove it while merging.

File: flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/PendingCheckpoint.java [hunk 2]
```diff
@@ -291,10 +292,8 @@ public CheckpointException getFailureCause() {
 	public CompletedCheckpoint finalizeCheckpoint() throws IOException {
 
 		synchronized (lock) {
-			checkState(areMasterStatesFullyAcknowledged(),
-				"Pending checkpoint has not been fully acknowledged by master states yet.");
-			checkState(areTasksFullyAcknowledged(),
-				"Pending checkpoint has not been fully acknowledged by tasks yet.");
+			checkState(!isDiscarded(), "checkpoint is discarded");
+			checkState(isFullyAcknowledged(), "Pending checkpoint has not been fully acknowledged yet");
```
[the remark this answers is not in the document]
Comment (StephanEwen) [reply flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/PendingCheckpoint.java]: True, but this is mainly a "sanity check" (guarding against accidental bugs) that should never actually trigger, so the error message detail is not so critical.
I found that previous changes (master hooks) were forgotten to add there, so this "check for all acks" seemed more safe/maintainable to me, compared to updating this whenever something would change.

File: flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/OperatorCoordinatorCheckpoints.java [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+		} catch (Exception e) {
+			throw new CompletionException(e);
+		}
+	}
+
+	// ------------------------------------------------------------------------
+
+	private static void acknowledgeAllCoordinators(PendingCheckpoint checkpoint, Collection<CoordinatorSnapshot> snapshots) {
+		for (final CoordinatorSnapshot snapshot : snapshots) {
+			checkpoint.acknowledgeCoordinatorState(snapshot.coordinator, snapshot.state);
```
[the remark this answers is not in the document]
Comment (StephanEwen) [reply flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/OperatorCoordinatorCheckpoints.java]: True, we can check for anything `!= SUCCESS` and exit early with an exception.

**Comment by StephanEwen** [comment]
Thank you for reviewing this @becketqin - I'll try to address the comments soon.

**Pull request closed without merging**